### ISSUE TITLE
xkb_keysym_from_name returns upper case keysym for case-insensitive match in a few cases

### ISSUE THREAD

**Issue by dottedmag** [xkbcommon/libxkbcommon#463]
`xkb_keysym_from_name` is documented to return a lowercase keysym if asked to do a case-insensitive match.

This fails (uppercase keysym is returned) for the following names:
```
dead_I
dead_i
dead_O
dead_o
dead_SCHWA
dead_schwa

kana_TSU
kana_tsu
kana_YA
kana_ya
kana_YO
kana_yo
```

**Comment by wismill** [comment]
Thanks for your feedback. This is a very interesting question indeed. Let’s look at the current doc (bold emphasis added):

> If you use the XKB_KEYSYM_CASE_INSENSITIVE flag and two keysym names
differ only by case, then **the lower-case keysym is returned**.  For
instance, for KEY_a and KEY_A, this function would return KEY_a for the
case-insensitive search.  If this functionality is needed, it is
recommended to first call this function without this flag; and if that
fails, only then to try with this flag, while possibly warning the user
he had misspelled the name, and might get wrong results.

So here, if two keysym names differ only by case, then the lower-case _keysym_ is returned, not the keysym corresponding to the lower-case keysym _name_. Indeed, `xkb_keysym_from_name` uses `xkb_keysym_is_lower` to test if a keysym is a lower-case keysym.

Let’s look at the example for keysyms `a` and `A`: we get the keysym `a` not because its name is lower case, but because `xkb_keysym_is_lower(XKB_KEY_a)` returns `true` and `xkb_keysym_is_lower(XKB_KEY_A)` returns `false`.

So the results are correct according to the doc:
- Katakana is _not a bicameral_ script, so e.g. `kana_tsu` is not the lower case of `kana_TSU`.
- As for the `dead_*` keysyms, they are not cased either because they do not correspond to characters.

But this also seems counter-intuitive to me: @bluetech, @whot is it intended? I think `xkb_keysym_is_lower` is not the right function to use, because I would expect to check only the name, not the corresponding character case:

```c
xkb_keysym_from_name("KANA_TSU", XKB_KEYSYM_CASE_INSENSITIVE) == XKB_KEY_kana_tsu;
xkb_keysym_from_name("dEaD_SCHWA", XKB_KEYSYM_CASE_INSENSITIVE) == XKB_KEY_dead_schwa;
```

_If_ the current behavior is wrong, I see two possible fixes:
1. Use a lookup table for _ambiguous_ entries, i.e. when we have multiple matches. This gives _exact results_ at the cost of a new LUT.
2. After the binary search in `xkb_keysym_from_name`, look for other keysyms with same case-folded name, compare their names and when keep the name that has a lower-case on the first character that differ. This may work but does not feel robust.

**Comment by dottedmag** [comment]
Indeed, the current behaviour is technically correct: documentation does not specify what happens if this function is called with a name that matches more that one keysym name and either none or more than one of those keysyms are lowercase.

However it's very surprising that `xkb_keysym_from_name("dead_o", 1) = XKB_KEY_dead_O` and `xkb_keysym_from_name("dead_a", 1) = XKB_KEY_dead_a`: any consistent behaviour would be better than current one.

**Comment by wismill** [comment]
> However it's very surprising that `xkb_keysym_from_name("dead_o", 1) = XKB_KEY_dead_O` and `xkb_keysym_from_name("dead_a", 1) = XKB_KEY_dead_a`: any consistent behaviour would be better than current one.

Ahah yes for this one you need to have [a look at the code](https://github.com/xkbcommon/libxkbcommon/blob/0d875ef56320e68b25416351dc397d59f2dd12cf/src/keysym.c#L288) to understand the output: it is due to the binary search matching (case insensitive) `dead_O` and then failing the `xkb_keysym_is_lower` test for the other matches (i.e. `dead_o`, defaulting to the keysym matched in the binary search.

**Comment by wismill** [comment]
I found an elegant fix for this issue: see #465.
